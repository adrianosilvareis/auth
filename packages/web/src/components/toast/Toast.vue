<template>
  <div class="toast" :class="{ show: show }" role="alert">
    <div class="toast-header">
      <strong class="me-auto">{{ title }}</strong>
      <small class="text-muted">{{ age }}</small>
      <btn-close @click="closed = !closed" />
    </div>
    <div class="toast-body">{{ message }}</div>
  </div>
</template>

<script lang="ts">
import { Options, Vue } from "vue-class-component";
import BtnClose from "@/components/btn/BtnClose.vue";
import CalcAgeUtil from "../../utils/methods/calc-age-util";

@Options({
  name: "Toast",
  components: {
    BtnClose,
  },
  props: {
    title: String,
    birthday: Date,
    message: String,
    duration: { type: Number, default: 20 },
  },
})
export default class Toast extends Vue {
  title!: string;
  birthday!: Date;
  message!: string;
  duration!: number;

  createdOn = new Date();
  now = new Date();
  closed = false;

  get show(): boolean {
    const time = this.createdOn.getTime();
    const validate = time + this.duration * 1000;
    const show = validate > this.now.getTime();
    return show && !this.closed;
  }

  get age(): string {
    const util = new CalcAgeUtil();
    util.age(this.createdOn, this.birthday);
    return util.age(this.createdOn, this.birthday);
  }

  mounted(): void {
    setInterval(() => {
      this.now = new Date();
    }, 1000);
  }
}
</script>

<style scoped></style>
