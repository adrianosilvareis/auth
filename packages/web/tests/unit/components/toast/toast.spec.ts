import { shallowMount } from "@vue/test-utils";
import Toast from "@/components/toast/Toast.vue";
import MockDate from "mockdate";

describe("Toast.vue", () => {
  beforeAll(() => {
    MockDate.set(new Date("2021-05-16T12:00:00.000Z"));
  });
  afterAll(() => {
    MockDate.reset();
  });
  describe("calcAge", () => {
    it("renders props values when passed", () => {
      const { wrapper } = makeWrapper();
      expect(wrapper.text()).toMatch("any_title");
      expect(wrapper.text()).toMatch("just now");
      expect(wrapper.text()).toMatch("any_message");
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders 'a minute ago' age", () => {
      const birthday = new Date("2021-05-16T11:59:00.000Z");
      const { wrapper } = makeWrapper(birthday);
      expect(wrapper.text()).toMatch("a minute ago");
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders '5 minutes ago' age", () => {
      const birthday = new Date("2021-05-16T11:55:00.000Z");
      const { wrapper } = makeWrapper(birthday);
      expect(wrapper.text()).toMatch("5 minutes ago");
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders 'a hour ago' age", () => {
      const birthday = new Date("2021-05-16T10:55:00.000Z");
      const { wrapper } = makeWrapper(birthday);
      expect(wrapper.text()).toMatch("a hour ago");
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders '7 hour ago' age", () => {
      const birthday = new Date("2021-05-16T04:55:00.000Z");
      const { wrapper } = makeWrapper(birthday);
      expect(wrapper.text()).toMatch("7 hours ago");
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders 'a day ago' age", () => {
      const birthday = new Date("2021-05-15T04:55:00.000Z");
      const { wrapper } = makeWrapper(birthday);
      expect(wrapper.text()).toMatch("a day ago");
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders '3 days ago' age", () => {
      const birthday = new Date("2021-05-13T04:55:00.000Z");
      const { wrapper } = makeWrapper(birthday);
      expect(wrapper.text()).toMatch("3 days ago");
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("show toast", () => {
    it("should render class show if createdOn + duration less than current date", async () => {
      const { wrapper } = makeWrapper();
      wrapper.vm.now = new Date("2021-05-16T12:00:00.000Z");
      expect(wrapper.html()).toMatch("show");
      expect(wrapper.element).toMatchSnapshot();
    });
    it("should not render class show if createdOn + duration greater than or equal current date", async () => {
      const { wrapper } = makeWrapper();
      wrapper.vm.now = new Date("2021-05-16T12:00:30.000Z");
      await wrapper.setData({}); // to update component
      expect(wrapper.vm.show).toBeFalsy();
      expect(wrapper.html()).not.toMatch("show");
      expect(wrapper.element).toMatchSnapshot();
    });
    it("should not render class show on click close button", async () => {
      const { wrapper } = makeWrapper();
      await wrapper.find("btn-close-stub").trigger("click");
      expect(wrapper.vm.show).toBeFalsy();
      expect(wrapper.html()).not.toMatch("show");
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});

function makeWrapper(
  birthday: Date = new Date(),
  title = "any_title",
  message = "any_message"
) {
  const wrapper = shallowMount(Toast, {
    props: { title, birthday, message },
  });

  return {
    wrapper,
  };
}
