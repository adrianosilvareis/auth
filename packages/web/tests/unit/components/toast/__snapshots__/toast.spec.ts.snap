// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast.vue renders '3 days ago' age 1`] = `
<div
  class="toast show"
  role="alert"
>
  <div
    class="toast-header"
  >
    <strong
      class="me-auto"
    >
      any_title
    </strong>
    <small
      class="text-muted"
    >
      3 days ago
    </small>
    <btn-close-stub />
  </div>
  <div
    class="toast-body"
  >
    any_message
  </div>
</div>
`;

exports[`Toast.vue renders '5 minutes ago' age 1`] = `
<div
  class="toast show"
  role="alert"
>
  <div
    class="toast-header"
  >
    <strong
      class="me-auto"
    >
      any_title
    </strong>
    <small
      class="text-muted"
    >
      5 minutes ago
    </small>
    <btn-close-stub />
  </div>
  <div
    class="toast-body"
  >
    any_message
  </div>
</div>
`;

exports[`Toast.vue renders '7 hour ago' age 1`] = `
<div
  class="toast show"
  role="alert"
>
  <div
    class="toast-header"
  >
    <strong
      class="me-auto"
    >
      any_title
    </strong>
    <small
      class="text-muted"
    >
      7 hours ago
    </small>
    <btn-close-stub />
  </div>
  <div
    class="toast-body"
  >
    any_message
  </div>
</div>
`;

exports[`Toast.vue renders 'a day ago' age 1`] = `
<div
  class="toast show"
  role="alert"
>
  <div
    class="toast-header"
  >
    <strong
      class="me-auto"
    >
      any_title
    </strong>
    <small
      class="text-muted"
    >
      a day ago
    </small>
    <btn-close-stub />
  </div>
  <div
    class="toast-body"
  >
    any_message
  </div>
</div>
`;

exports[`Toast.vue renders 'a hour ago' age 1`] = `
<div
  class="toast show"
  role="alert"
>
  <div
    class="toast-header"
  >
    <strong
      class="me-auto"
    >
      any_title
    </strong>
    <small
      class="text-muted"
    >
      a hour ago
    </small>
    <btn-close-stub />
  </div>
  <div
    class="toast-body"
  >
    any_message
  </div>
</div>
`;

exports[`Toast.vue renders 'a minute ago' age 1`] = `
<div
  class="toast show"
  role="alert"
>
  <div
    class="toast-header"
  >
    <strong
      class="me-auto"
    >
      any_title
    </strong>
    <small
      class="text-muted"
    >
      a minute ago
    </small>
    <btn-close-stub />
  </div>
  <div
    class="toast-body"
  >
    any_message
  </div>
</div>
`;

exports[`Toast.vue renders props values when passed 1`] = `
<div
  class="toast show"
  role="alert"
>
  <div
    class="toast-header"
  >
    <strong
      class="me-auto"
    >
      any_title
    </strong>
    <small
      class="text-muted"
    >
      just now
    </small>
    <btn-close-stub />
  </div>
  <div
    class="toast-body"
  >
    any_message
  </div>
</div>
`;
